firstapp/app/src/androidTest/java/com/example/firstapp/ShellUtility.java

```diff
@@ -287,29 +330,151 @@ public Action parseStringAction(String str, int idx) throws invalidInputExceptio
+        }
+
+        //Log average action durations
+        long[] averageActionDurations = averageColumns(allActionDurations);
+        Log.i("averages-actions", "AVERAGE:     " + Arrays.toString(averageActionDurations) + ", TOTAL: " + sumArr(averageActionDurations));
+
+
+        //Log time stamps relative to moment first measured action became available and store iteration durations
+        long[][] allRelativeStamps = new long[iterations][];
+        long[][] iterDurations = new long[iterations][2]; //to be used later for finding median
+        for (int iter = 0; iter < iterations; iter++) {
+            long[] actionStamps = allActionStamps[iter];
+            long[] relativeStamps = relativeValues(actionStamps);
+
+            allRelativeStamps[iter] = relativeStamps;
+            iterDurations[iter][0] = iter;
+            iterDurations[iter][1] = relativeStamps[relativeStamps.length - 1];
+            Log.i("iterations-stamps", "ITERATION " + (iter + 2) + ": " + Arrays.toString(relativeStamps));
+        }
+
+        //log average relative stamps
+        long[] averageRelativeStamps = averageColumns(allRelativeStamps);
+        Log.i("averages-stamps", "AVERAGE:     " + Arrays.toString(averageRelativeStamps));
+
+        //log median iteration
+        Arrays.sort(iterDurations, (a,b) -> Long.compare(a[1], b[1]));
+        int median_idx = (int) iterDurations[iterDurations.length / 2][0];
+        long med_start = allActionStamps[median_idx][0];
+        long med_end = allActionStamps[median_idx][allActionStamps[0].length - 1];
+        Log.i("median", "MEDIAN RUN: " + (median_idx + 2));
+        //Log.i("median", "MEDIAN RUN STARTS @: " + miliseconds_to_time(med_start));
+        //Log.i("median", "MEDIAN RUN ENDS @: " + miliseconds_to_time(med_end));
+
+        //log median clip data
+        Log.i("clip_start", "" + med_start);
+        Log.i("clip_end", "" + med_end);
+    }
+
+    /******************************************************************************
+     * Executes a CUJ with preparatory actions pre and measured actions post.
+     * Caches data and then runs through the CUJ iterations times and logs results
+     ******************************************************************************/
+    public void executeCUJ(String[] preCUJ, String[] postCUJ, int iterations) throws Exception {
+        if (postCUJ.length < 2) throw new invalidInputException("Measured CUJ must have length >= 2. Make sure you have a final 'dummy' action");
         String[] cujStrings = new String[preCUJ.length + postCUJ.length];
         System.arraycopy(preCUJ, 0, cujStrings, 0, preCUJ.length);
         System.arraycopy(postCUJ, 0, cujStrings, preCUJ.length, postCUJ.length);
         Action[] cuj = parseStringCUJ(cujStrings);
-        UiObject[] cachedObjects = cacheCUJ(cuj);
-    }
 
+        //caching run
+        cacheCUJ(cuj);
```

Comment:
Ideally don't want non-CUJ actions in the same video.

Reply:
the script will use logcat information to clip out the video we want. I'm not sure whether there's an easy way to avoid recording this part at all without starting the recording within the test or storing cached data to a file and passing that into a new test. (which might be worth it if this inclusion is hurting us due to the 3 minute max recording length.)